File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/DelayHandle/DelayNoOrPriorityHandleMgrBase.cpp
#include "MyProject.h"
#include "DelayNoOrPriorityHandleMgrBase.h"

MY_BEGIN_NAMESPACE(MyNS)

DelayNoOrPriorityHandleMgrBase::DelayNoOrPriorityHandleMgrBase()
{
	this.mIsDispose = false;
	this.mLoopDepth = new LoopDepth();
	this.mLoopDepth.setZeroHandle(null, this._processDelayObjects, 0);

	if(MacroDef.DEBUG_SYS)
	{
		this.mDebugUniqueId = (int)Ctx.msInstance.mDebugSys.mDebugPriorityHandleUniqueId.genNewId();
	}
}

virtual void DelayNoOrPriorityHandleMgrBase::init()
{

}

virtual void DelayNoOrPriorityHandleMgrBase::dispose()
{
	this.mIsDispose = true;

	if (null != this.mDeferredAddQueue)
	{
		this.mDeferredAddQueue.dispose();
		this.mDeferredAddQueue = null;
	}
	if (null != this.mDeferredDelQueue)
	{
		this.mDeferredDelQueue.dispose();
		this.mDeferredDelQueue = null;
	}
	if (null != this.mLoopDepth)
	{
		this.mLoopDepth.dispose();
		this.mLoopDepth = null;
	}
}

int DelayNoOrPriorityHandleMgrBase::getDebugUniqueId()
{
	return this.mDebugUniqueId;
}

void DelayNoOrPriorityHandleMgrBase::_addObject(IDelayHandleItem delayObject, float priority)
{
	if(this.mLoopDepth._isInDepth())
	{
		// 如果添加列表中没有
		if (!this.mDeferredAddQueue.contains(delayObject as INoOrPriorityObject))
		{
			// 如果已经添加到删除列表中
			if (this.mDeferredDelQueue.contains(delayObject as INoOrPriorityObject))
			{
				this.mDeferredDelQueue.removeNoOrPriorityObject(delayObject as INoOrPriorityObject);
			}
			
			this.mDeferredAddQueue.addNoOrPriorityObject(delayObject as INoOrPriorityObject, priority);
		}
	}
}

void DelayNoOrPriorityHandleMgrBase::_removeObject(IDelayHandleItem delayObject)
{
	if (delayObject is INoOrPriorityObject)
	{
		if (this.mLoopDepth._isInDepth())
		{
			if (!this.mDeferredDelQueue.contains(delayObject as INoOrPriorityObject))
			{
				if (this.mDeferredAddQueue.contains(delayObject as INoOrPriorityObject))    // 如果已经添加到删除列表中
				{
					this.mDeferredAddQueue.removeNoOrPriorityObject(delayObject as INoOrPriorityObject);
				}

				delayObject.setClientDispose(true);

				this.mDeferredDelQueue.addNoOrPriorityObject(delayObject as INoOrPriorityObject);
			}
		}
	}
	else
	{
		Ctx.msInstance.mLogSys.log("delayObject is not INoOrPriorityObject", LogTypeId.eLogDataStruct);
	}
}

// 只要调用会添加或者删除列表元素，就需要调用这个接口
void DelayNoOrPriorityHandleMgrBase::_incDepth()
{
	this.mLoopDepth._incDepth();
}

// 只要调用会添加或者删除列表元素，就需要调用这个接口
void DelayNoOrPriorityHandleMgrBase::_decDepth()
{
	if (null != this.mLoopDepth)
	{
		this.mLoopDepth._decDepth();
	}
	else
	{
		// 已经释放了资源，不用再处理了
	}
}

bool DelayNoOrPriorityHandleMgrBase::_isInDepth()
{
	bool ret = false;

	if (null != this.mLoopDepth)
	{
		ret = this.mLoopDepth._isInDepth();
	}
	else
	{
		// 资源已经释放
	}

	return ret;
}

void DelayNoOrPriorityHandleMgrBase::_processDelayObjects(IDispatchObject dispObj, uint eventId)
{
	int idx = 0;
	// len 是 Python 的关键字
	int elemLen = 0;

	if (!this.mLoopDepth._isInDepth())       // 只有全部退出循环后，才能处理添加删除
	{
		if (this.mDeferredAddQueue.count() > 0)
		{
			idx = 0;
			elemLen = this.mDeferredAddQueue.count();

			while(idx < elemLen)
			{
				this._addObject(this.mDeferredAddQueue.get(idx) as IDelayHandleItem);

				idx += 1;
			}

			this.mDeferredAddQueue.clear();
		}

		if (this.mDeferredDelQueue.count() > 0)
		{
			idx = 0;
			elemLen = this.mDeferredDelQueue.count();

			while(idx < elemLen)
			{
				this._removeObject(this.mDeferredDelQueue.get(idx) as IDelayHandleItem);

				idx += 1;
			}

			this.mDeferredDelQueue.clear();
		}
	}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/DelayHandle/DelayNoOrPriorityHandleMgrBase.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

-- 当需要管理的对象可能在遍历中间添加的时候，需要这个管理器
local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "DelayNoOrPriorityHandleMgrBase";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.mLoopDepth = GlobalNS.new(GlobalNS.LoopDepth); 	-- 是否在循环中，支持多层嵌套，就是循环中再次调用循环
	self.mLoopDepth:setZeroHandle(self, M.processDelayObjects);
end

function M:init()

end

function M:dispose()

end

function M:addObject(delayObject, priority)
	if(self.mLoopDepth:isInDepth()) then
		if (not self.mDeferredAddQueue:contains(delayObject)) then       -- 如果添加列表中没有
			if (self.mDeferredDelQueue:contains(delayObject)) then    -- 如果已经添加到删除列表中
				self.mDeferredDelQueue.removeNoOrPriorityObject(delayObject);
			end
			
			self.mDeferredAddQueue:addNoOrPriorityObject(delayObject);
		end
	end
end

function M:removeObject(delayObject)
	if (self.mLoopDepth:isInDepth()) then
		if (not self.mDeferredDelQueue:contains(delayObject)) then
			if (self.mDeferredAddQueue:contains(delayObject)) then   -- 如果已经添加到删除列表中
				self.mDeferredAddQueue:removeNoOrPriorityObject(delayObject);
			end

			delayObject:setClientDispose(true);
			
			self.mDeferredDelQueue:addNoOrPriorityObject(delayObject);
		end
	end
end

function M:processDelayObjects()
	local idx = 0;
	-- len 是 Python 的关键字
	local elemLen = 0;

	if (not self.mLoopDepth:isInDepth()) then      -- 只有全部退出循环后，才能处理添加删除
		if (self.mDeferredAddQueue:count() > 0) then
			idx = 0;
			elemLen = self.mDeferredAddQueue:count();

			while(idx < elemLen) do
				self:addObject(self.mDeferredAddQueue:get(idx));

				idx = idx + 1;
			end

			self.mDeferredAddQueue:clear();
		end

		if (self.mDeferredDelQueue:count() > 0) then
			idx = 0;
			elemLen = self.mDeferredDelQueue:count();

			while(idx < elemLen) do
				self:removeObject(self.mDeferredDelQueue:get(idx));

				idx = idx + 1;
			end

			self.mDeferredDelQueue:clear();
		end
	end
end

function M:incDepth()
	self.mLoopDepth:incDepth();
end

function M:decDepth()
	self.mLoopDepth:decDepth();
end

function M:isInDepth()
	return self.mLoopDepth:isInDepth();
end

return M;